alibi_detect/saving/_pytorch/saving.py

```diff
@@ -0,0 +1,134 @@
+
+
+def save_model_config(model: Callable,
+                      base_path: Path,
+                      input_shape: Optional[tuple],
+                      local_path: Path = Path('.')) -> Tuple[dict, Optional[dict]]:
+    """
+    Save a PyTorch model to a config dictionary. When a model has a text embedding model contained within it,
+    this is extracted and saved separately.
+
+    Parameters
+    ----------
+    model
+        The model to save.
+    base_path
+        Base filepath to save to (the location of the `config.toml` file).
+    input_shape
+        The input dimensions of the model (after the optional embedding has been applied).
+    local_path
+        A local (relative) filepath to append to base_path.
+
+    Returns
+    -------
+    A tuple containing the model and embedding config dicts.
+    """
+    cfg_model = None  # type: Optional[Dict[str, Any]]
+    cfg_embed = None  # type: Optional[Dict[str, Any]]
+    if isinstance(model, UAE):
+        layers = list(model.encoder.children())
+        if isinstance(layers[0], TransformerEmbedding):  # if UAE contains embedding and encoder
+            # embedding
+            embed = layers[0]
+            cfg_embed = save_embedding_config(embed, base_path, local_path.joinpath('embedding'))
+            # preprocessing encoder
+            model = layers[1]
+        else:  # If UAE is simply an encoder
+            model = model.encoder
+    elif isinstance(model, TransformerEmbedding):
+        cfg_embed = save_embedding_config(model, base_path, local_path.joinpath('embedding'))
+        model = None
+    elif isinstance(model, HiddenOutput):
+        model = model.model
+    elif isinstance(model, nn.Module):  # Last as TransferEmbedding and UAE are nn.Module's
+        model = model
+    else:
+        raise ValueError('Model not recognised, cannot save.')
+
+    if model is not None:
+        filepath = base_path.joinpath(local_path)
+        save_model(model, filepath=filepath)
+        cfg_model = {
+            'flavour': Framework.PYTORCH.value,
+            'src': local_path.joinpath('model')
+        }
+    return cfg_model, cfg_embed
+
+
+def save_model(model: Union[nn.Module, nn.Sequential],
+               filepath: Union[str, os.PathLike],
+               save_dir: Union[str, os.PathLike] = 'model') -> None:
+    """
+    Save PyTorch model.
+
+    Parameters
+    ----------
+    model
+        The PyTorch model to save.
+    filepath
+        Save directory.
+     save_dir
+        Name of folder to save to within the filepath directory.
+    """
+    # create folder to save model in
+    model_path = Path(filepath).joinpath(save_dir)
+    if not model_path.is_dir():
+        logger.warning('Directory {} does not exist and is now created.'.format(model_path))
+        model_path.mkdir(parents=True, exist_ok=True)
+
+    # save model
+    model_path = model_path.joinpath('model.pt')
+
+    if isinstance(model, (nn.Module, nn.Sequential)):
```

Comment:
Do we need to have both types here?

Reply:
Thanks. Have removed for now (pending discussion in #668 )